### ISSUE TITLE
Ship default abbreviations

### ISSUE THREAD

**Issue by faho** [fish-shell/fish-shell#9462]
I think everyone's enjoying the new expanded abbreviations so far.

It might be cool if we shipped some and turned them on by default.

Some possibilities include the two examples from our CHANGELOG:

- The ever-popular `!!` and `!$` (I wouldn't add a full history-expansion thing)
- The `...` multi-cd thing

But there are more. Here's some I built:

- Turn any directory used as a command into a `cd` - this would allow removing implicit cd as part of the language (which never sat well with me)
- Turn `=foo` into the path to `foo` (i.e. execute `command -s foo` implicitly) if that's a command - this is *syntax* in zsh
- Add a pipe to a pager, *with added colorisation* - turn `grep foo pag ` into `grep foo --color=auto | less -R` and `gcc pag` into `gcc -fdiagnostics-color=always | less -R` - this would replace `__fish_paginate` (and would be cleaner with #9412 - otherwise you end up removing decorating keywords manually)

And here are functions we ship that could be an abbreviation instead (which means it removes buffering!):

- ll and la - cheesy `ls` aliases (frankly I would be okay with removing these entirely)
- `ls`, `grep`, `diff` - these just add color

----

Of course the big question is *how* we ship these. In particular they would need a way to turn them off, ideally individually.

Since we don't have autoloading for abbrs, the best we could do would be a conf.d file, which allows both adding a later file to `abbr --erase` select ones as well as shadowing the entire file at once.

Plus, this needs extremely clear communication on how to *skip* the expansion. The user didn't opt-in to these, so they might not even know they exist. (this is part of why I've been so adamant about not doing quiet abbreviations and always documenting ctrl-space ctrl-space ctrl-space everywhere)

The alternative is to ship them and don't turn them on - add them in the docs, ask people to run `fish_cool_abbrs` on startup? We don't typically do that but this is a tricky subject.

Or we could use our blog and show some off.

This is very much a "request for comments" - I'm not set on doing this, and especially not any particular example. I just think it's worth thinking about.

**Comment by 0ion9** [comment]
> * Turn `=foo` into the path to `foo` (i.e. execute `command -s foo` implicitly) if that's a command - this is _syntax_ in zsh

Syntax is better. In terms of not producing a screwed up history.

Not to say that this stuff isn't cool, but the way that abbrs (other than typo corrections) interfere with the usability of history search says to me that they should be changed, rather than promoted. Maybe that's totally infeasible but that is still my opinion.

(if you started a command by typing `xyz `, then history searching `xyz ` should find it. This is the case for everything other than abbrs. Functions are an abstraction which the user can usually safely forget the content of. Abbreviations are an abstraction which the user is required to know the content of in order to search them (ie. a leaky abstraction))

Of the examples you present here, `!!` is both popular and doesn't have any real searchability problems, so I guess I can +1 that, at least.

**Comment by faho** [comment]
>Syntax is better. In terms of not producing a screwed up history.

Adding syntax for a silly shortcut like this isn't something we do.

And there is no real way to produce a "screwed up history", because it tells you what you are executing - it just inserts the thing.

Of course if you want the path to the command to track where it's currently is, go ahead and use `(command -s foo)`, which will expand right then.

>if you started a command by typing xyz , then history searching xyz should find it

But that *isn't* the command you executed.

>Abbreviations are an abstraction which the user is required to know the content of in order to search them (ie. a leaky abstraction))

And it leaks right there, immediately, after pressing space, into your commandline. I'd call it "transparent" instead of "leaky".

----

Anyway, that's not what I'm requesting comments about. This isn't about abbreviations as a general concept, this is about how much we want to enable it *by default*. And we don't want to replace any real functions with it, but when it comes to a cheesy `ls --color=auto` thing or bindings that insert `| less` (where you *also* can't search for "alt+p") I can't help but see them as better.

Abbreviations are still optional - if you don't like them you can entirely ignore them and use functions instead. That is why I'm asking for an easy way to opt-out.

**Comment by 0ion9** [comment]
> Adding syntax for a silly shortcut like this isn't something we do.

And I'm not asking you to -- obscure syntax is not good UI either.

I'm asking you not to take the confusing UI fish had in this regard for a long time (since abbrs and history were both a thing?) and  show publicly how to make it more confusing, while presenting that as a feature.

> Of course if you want the path to the command to track where it's currently is, then [..]

My opinion was that there was no case in which having the full path to a command in your history would be useful. But I've thought about what you said and I can see that it would have some use with virtual environments / chroots.


> But that _isn't_ the command you executed.

It's not what *fish executed*, but it *is* the mnemonic that abbrs train me to use. Especially so in cases where there is no space (see below). 

> And it leaks right there, immediately, after pressing space, into your commandline. I'd call it "transparent" instead of "leaky".

With space, yes, the user gets to see it before execution, and that is less confusing.

Let's review the examples you provided:

* *Multicd*: Only token, so no space unless for some reason you like to add one
* *Directory used as command*:  Only token
* `!!`: Depends on whether you are trying to repeat it or edit it. I'd say the classic example is `sudo !!`
* `=foo`:  Either the only token (`=foo` execute with full path) or last token (`set foobinary =foo`), AFAICS
* *Pipe to a pager* : Last token unless you want to give options to your pager

Without space, the user sees the command only after execution. 

> We are _extremely_ happy with how abbrs currently work and are _not_ about to change them to be less powerful again.

I don't see how 'how abbrs currently work' is relevant. This problem existed before, with their less powerful form, and it exists now: the point is that most of the examples provided would be a further increment to confusing user interactions.

(the same problem exists in BASH, ZSH, etc, AFAIK. In every case it's not abbr itself per se, it's **the entire chain abbr -> keyboard input -> history tracking** that is, in sum, an interface screw.)

> This isn't about abbreviations as a general concept, this is about how much we want to enable it by default. And we don't want to replace any real functions with it, but when it comes to a cheesy ls --color=auto thing or bindings that insert | less (where you also can't search for "alt+p") I can't help but see them as better.

I tried to address that a bit (`!!`), but I guess I can also say, the `ls` abbreviations don't seem like a big deal to include by default.
If `pag` expands to the value of $PAGER, which I guess it probably should, then that's on the user mostly whether it makes a mess or not.

**Comment by faho** [comment]
This isn't the issue to air all your gripes with abbreviations. If you have a concrete idea on how to improve the abbreviation system, or a concrete problem with it, please open another issue and ask for that. But: "Remove abbrs" or "Revert the recent abbr expansion" aren't on the table.

And I will point out again that abbrs are optional. If you don't like them, or don't like a particular one, you always have the option of not using them.

>Without space, the user sees the command only after execution.

Abbreviations are currently expanded for the bind functions "expand-abbr" (which space does by default) and "execute". This is also how pressing ctrl-space lets you skip an abbreviation (and leave the key in literally).

Removing it from the latter is still on my agenda. In that case it would be possible to bind return to "expand-abbr or execute", so if it expands it doesn't execute. We might do that by default.

**Comment by faho** [comment]
>If pag expands to the value of $PAGER

We would try to use $PAGER, yes - for less we would have to add `-R`, for lv we need `-c`, other pagers might need similar options. Whether we either fall back to less (if available) or refuse to expand (/expand to nothing) or just use $PAGER if we don't know it is something we'd have to figure out.

Really the advantage here is automatic colorization, which *can't* be done with a function since it needs to touch both ends of the pipe. And we would move the cursor back to the actual command, which is annoying with a binding.

**Comment by ridiculousfish** [comment]
As for shipping default abbreviations, one approach is to follow the model set by `fish_mode_prompt`. We could have a function `fish_default_abbrs` where we add default abbreviations. Users can disable the default by defining `fish_default_abbrs` to be empty.

If we define default abbrs then I think we should try to stick to special characters. Multi-cd and history expansion seem good, `=foo` seems OK, I'm not sold on `grep` and `ls`.  fish's wrapping of `grep` and `ls` into functions is IMO a mistake in hindsight, and injecting characters into the command line that the user didn't type I think would make it worse.

I'd be happy to remove `ll` though surely some users will miss it.

**Comment by faho** [comment]
>We could have a function fish_default_abbrs where we add default abbreviations. Users can disable the default by defining fish_default_abbrs to be empty.

I think I would call it just `fish_abbrs` - the "default" makes sense if you just think of using it all-or-nothing, but if you want to disable just one specific abbr you'd `funced` it.

You could also use `abbr --erase`, but that is then awkward with timing - at which point do we set up the default abbreviations? It would fit neatly into __fish_config_interactive, but that's executed after config.fish.

----

>Multi-cd and history expansion seem good, =foo seems OK

Yeah I think I'm most interested in multi-cd and `!!`. The `=foo` thing is a silly example to show how this can be used in lieu of syntax.

>fish's wrapping of grep and ls into functions is IMO a mistake in hindsight

Yup. My thought here was that we'd *show* people what they're executing (and of course removing the function means we get to sidestep buffering and stuff).

How well that works relies on how well we can teach people to skip or disable the abbreviation as they want.

Or we could leave those in the documentation or a separate "fish_color_abbrs" that's off-by-default.

(really, ideally these tools would just grow better defaults, but this is unix and we don't change things)

>I'd be happy to remove ll though surely some users will miss it.

It's one line for something I don't think most people use. It's weird that we ship these - ll is `ls -lh` and la is `ls -lAh`.

We don't do that for anything else and I have never used them.

**Comment by exploide** [comment]
Since this issue requests for comments, these are my two cents.

`...` looks handy, don't know if I would get used to it but sounds cool. Concerning `!!` I was always on your side @faho when you explained people that <kbd>Alt</kbd>+<kbd>S</kbd> is even less to type. But given that a lot of people seem to like it, maybe it's good.

Regarding coloring of `grep` or `ls` you got me when mentioning that this solves the buffering issue. But I'm totally with @ridiculousfish that "injecting characters into the command line that the user didn't type" is something I would really dislike. First I read in #9313 that the new abbr could have a `--quiet`, which would workaround this problem, but I noticed that options has been dropped.

> I'd be happy to remove ll though surely some users will miss it.

I would miss them totally. I never ever use plain `ls`, I always do `ll`. If I enter a shell where this isn't available, the first thing I do is setting an alias. But it is also pretty common to have it. Many default `.bashrc` files ship this alias, sometimes commented out. Even on the ash on my OpenWrt router, there is a `ll` alias by default. Given that fish is the user-friendly shell I would personally expect that alias (or function here).

**Comment by faho** [comment]
>First I read in https://github.com/fish-shell/fish-shell/pull/9313 that the new abbr could have a --quiet, which would workaround this problem, but I noticed that option has been dropped.

It has, because it's confusing - now we're back to it executing something different from what was typed except not even `type` will tell you what that was, and neither `command foo` nor ctrl-space are guaranteed to skip it. That's the worst of both worlds.

`grep --color=auto` as a function can be skipped with `command grep`. `grep --color=auto` as an abbreviation can be skipped by typing `grep` and pressing ctrl-space.

The big advantage of the latter is that you basically don't *have* to? Like I said, `--color=auto` could just be the default, there's basically no need to skip it (the exceptions are if grep doesn't support it or if your terminal doesn't support color, both of which we could just detect).

**Comment by krobelus** [comment]
I think `!!` expanding on Enter without confirmation is a bit scary but I'm sure it's fine in practice.

> You could also use `abbr --erase`, but that is then awkward with timing - at which point do we set up the default abbreviations? It would fit neatly into __fish_config_interactive, but that's executed after config.fish.

I'd prefer this because it's the obvious way to disable individual default abbreviation. They don't need to be in `__fish_config_interactive`.
`bind` solves this problem with `bind --preset`, which might be interesting here too if we want this kind of introspectability.

> >I'd be happy to remove ll though surely some users will miss it.
> It's one line for something I don't think most people use. It's weird that we ship these - ll is `ls -lh` and la is `ls -lAh`.
> We don't do that for anything else and I have never used them.

I use `ll` all the time and as mentioned above some distros add this alias to their default shell.

**Comment by exploide** [comment]
> The big advantage of the latter is that you basically don't have to? Like I said, --color=auto could just be the default, there's basically no need to skip it (the exceptions are if grep doesn't support it or if your terminal doesn't support color, both of which we could just detect).

Yes, after submitting my comment I thought maybe I shouldn't stick that much with the behavior I'm familiar with and just let you do whatever you plan and see how to use it.
But I still have doubts concerning expansion of stuff like color flags within my commandline. This make everything longer and less overseeable. I often edit the previous commandline, e.g. of a longer pipe, to run it again afterwards. And the more stuff there is, the longer it takes to navigate around the correct spot.

And I also feel, a function, as it currently is, could be the more appropriate place to fix unreasonable defaults of commands like grep. Isn't the only drawback of the function that is has the buffering issue? Is the approach of fixing this with multithreading off the table for the foreseeable future?

**Comment by faho** [comment]
>I think !! expanding on Enter without confirmation is a bit scary but I'm sure it's fine in practice.

Note: I have been saying for years that "execute" shouldn't expand abbreviations and immediately execute things, every chance I get. More people than I would expect have been against this - see e.g. #8423.

> They don't need to be in __fish_config_interactive.

Okay, if they aren't where are they? Should they just be set up in share/config.fish?

>And I also feel, a function, as it currently is, could be the more appropriate place to fix unreasonable defaults of commands like grep

It has the downside of buffering, but there's also just the thing that... you're not running `grep`. It looks like you are running `grep` but you aren't, and so you run `grep` on another machine and it somehow behaves differently (even if very slightly).

So *showing* people that "hey, this needs --color=auto to behave sensibly" feels more open and honest.

>Is the approach of fixing this with multithreading off the table for the foreseeable future?

It's not a short-term thing, no.

**Comment by septatrix** [comment]
Just my two cents regarding `ll`, `grep`, `ls` and `diff` as abbreviations:
Having them auto expand would likely add a lot of noise to the prompt and history. `ll` is a lot cleaner than `ls --color=auto -lh` sprinkled throughout the terminal scrollback and history.

The other suggestions like `!!` however seem like a great way to make the transition easier for people coming from bash/zsh without full fledged history expansion). `...` also seems like a neat shortcut for some people (though I would probably stay with `../..` et al as I often use tab to check how much I already traveled up in the tree).

**Comment by faho** [comment]
>Could you please clarify what buffering issue you are referring to for those with less in-depth knowledge about fish?

The problem, in a nutshell, is this:

Fish runs the actual script in a single thread. It can only ever run one builtin/function at the same time.

So when you do, e.g. `string | string`, that must complete the first `string` before the second can run. This also goes for function calls, so this hits with

```fish
yes | grep y | grep y
```

if `grep` is a function. You'll basically see it hang because the first `grep` never stops, because `yes` doesn't.

The easiest thing to do here is to simply remove the function. Since abbrs disappear after they're expanded, that's a nice solution.

(other shells solve this issue with "subshells", they fork off for any process in the middle of a pipeline, but that introduces awkward semantics like not being able to set variables, and *most* of the time this is fairly benign. our preferred solution here is multi-threaded execution, but to do that with confidence we need language support - hence #9512)

**Comment by kwstannard** [comment]
I can open a new issue for this, but `!!` is not the same thing as `$history[1]`.

```
#bash 
$ echo hello
$ VAR=1 !!
hello

#fish
$ echo hello
$ VAR=1 $history[1]
fish: Unknown command: 'echo hello'
```

**Comment by faho** [comment]
>I can open a new issue for this, but !! is not the same thing as $history[1].

@kwstannard That is working as expected and not an issue - `$history[1]` is the entire last history element *as one thing*, without being split. You're doing the equivalent of running `VAR=1 'echo hello'`. Fish fundamentally *does not split variables after they have been set*. This is unlike bash, which does what it calls "word splitting" - which is why you should (almost) always double-quote bash variables.

However, with the abbreviation, what this is about, it works. Like [the release notes explain](https://fishshell.com/docs/current/relnotes.html#id1):

```fish
function last_history_item; echo $history[1]; end
abbr -a !! --position anywhere --function last_history_item
```

This replaces `!!`, in the commandline, with the literal value of $history[1], and so you end up with:

```fish
VAR=1 !!
```

turning into

```fish
VAR=1 echo hello
```

Because there's that decoupling that will work.

---

(there is one tiny issue in that it should be `echo -- $history[1]` because otherwise echo will complain if $history[1] starts with a `-`)

**Comment by kwstannard** [comment]
nice! Thanks

**Comment by julian-a-avar-c** [comment]
Hello! Beginner here, just went into a loop because I found `!!` in a tutorial and got super confused. A friend recommended fish and I'm starting to realize that it's more than just `.bashrc` that changes.

**Comment by lnicola** [comment]
Hello. Just gave this a try:

```fish
$ function _which
   which $argv
end
$ abbr --add = --position anywhere --function _which
$ _which ls
/usr/bin/ls
$ echo =ls
=ls
```

Is there a way to make this work?

**Comment by krobelus** [comment]
> $ abbr --add = --position anywhere --function _which
> $ _which ls
> /usr/bin/ls
> $ echo =ls
> =ls

needs "--regex '=\w+'", else it will match only literal =

**Comment by lnicola** [comment]
Thanks. That also needs something like `which $(string sub --start=2 $argv)` because `$argv` includes the `=`.

**Comment by kpbaks** [comment]
> Thanks. That also needs something like `which $(string sub --start=2 $argv)` because `$argv` includes the `=`.

Instead of relying on the `which` binary, you can use the builtin `command --search`

```fish
command --search (string sub --start=2 $argv)
```
